client/utils/language-utils.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * Utility functions for language detection and handling
+ */
+
+function getPreferredLanguage(
+  supportedLanguages: string[] = [],
+  defaultLanguage: string = 'en'
+): string | undefined {
+  if (typeof navigator === 'undefined') {
+    return defaultLanguage;
+  }
+
+  const normalizeLanguage = (langCode: string) => langCode.toLowerCase().trim();
+  const normalizedSupported = supportedLanguages.map(normalizeLanguage);
+
+  /**
+   * Attempts to find a match in normalizedSupported given a browser-provided language.
+   * Prioritizes exact match of both language and region (eg. 'en-GB'), falls back to base-language match (eg. 'en').
+   */
+  function findMatch(inputLang: string): string | undefined {
+    const normalizedLang = normalizeLanguage(inputLang);
+
+    const exactMatchIndex = normalizedSupported.indexOf(normalizedLang);
+    if (exactMatchIndex !== -1) return supportedLanguages[exactMatchIndex];
+
+    const baseLanguage = normalizedLang.split('-')[0];
+    const partialMatchIndex = normalizedSupported.findIndex(
+      (lang) => lang === baseLanguage || lang.startsWith(`${baseLanguage}-`)
+    );
+    if (partialMatchIndex !== -1) return supportedLanguages[partialMatchIndex];
+
+    // eslint-disable-next-line consistent-return
+    return undefined;
+  }
+
+  // Try navigator.languages list first
+  if (Array.isArray(navigator.languages)) {
+    for (let i = 0; i < navigator.languages.length; i++) {
+      const match = findMatch(navigator.languages[i]);
+      if (match) return match;
+    }
+  }
+
+  // Fallback to navigator.language
+  if (navigator.language) {
+    const match = findMatch(navigator.language);
+    if (match) return match;
+  }
+
+  return defaultLanguage;
+}
+
+export default getPreferredLanguage;
```

Comment:
can we put the default export on the function itself here too?

Reply:
or even better... maybe we can make it a named export instead

https://google.github.io/styleguide/tsguide.html#exports